File: src/bmsx/render/shared/hardware/lighting.ts
import { ensureDirectionalLightRecord, ensurePointLightRecord, type AmbientLight, type DirectionalLight, type PointLight } from '../../3d/light';

const ambientLights = new Map<string, AmbientLight>();
const directionalLights = new Map<string, DirectionalLight>();
const pointLights = new Map<string, PointLight>();
let hardwareLightingDirty = false;

function writeVec3(target: [number, number, number], source: readonly number[]): void {
	target[0] = source[0];
	target[1] = source[1];
	target[2] = source[2];
}

function ambientLightRecord(id: string): AmbientLight {
	let record = ambientLights.get(id);
	if (!record) {
		record = { type: 'ambient', color: [0, 0, 0], intensity: 0 };
		ambientLights.set(id, record);
	}
	return record;
}

export function putHardwareAmbientLight(id: string, light: AmbientLight): void {
	const record = ambientLightRecord(id);
	writeVec3(record.color, light.color);
	record.intensity = light.intensity;
	hardwareLightingDirty = true;
}

export function putHardwareDirectionalLight(id: string, light: DirectionalLight): void {
	const record = ensureDirectionalLightRecord(directionalLights, id);
	writeVec3(record.color, light.color);
	writeVec3(record.orientation, light.orientation);
	record.intensity = light.intensity;
	hardwareLightingDirty = true;
}

export function putHardwarePointLight(id: string, light: PointLight): void {
	const record = ensurePointLightRecord(pointLights, id);
	writeVec3(record.color, light.color);
	writeVec3(record.pos, light.pos);
	record.range = light.range;
	record.intensity = light.intensity;
	hardwareLightingDirty = true;
}

export function clearHardwareLighting(): void {
	const hadLights = ambientLights.size !== 0 || directionalLights.size !== 0 || pointLights.size !== 0;
	ambientLights.clear();
	directionalLights.clear();
	pointLights.clear();
	if (hadLights) {
		hardwareLightingDirty = true;
	}
}

export function consumeHardwareLightingDirty(): boolean {
	if (!hardwareLightingDirty) {
		return false;
	}
	hardwareLightingDirty = false;
	return true;
}

export function getHardwareAmbientLights(): ReadonlyMap<string, AmbientLight> {
	return ambientLights;
}

export function getHardwareDirectionalLights(): ReadonlyMap<string, DirectionalLight> {
	return directionalLights;
}

export function getHardwarePointLights(): ReadonlyMap<string, PointLight> {
	return pointLights;
}

export function resolveHardwareAmbientLight(): AmbientLight | null {
	if (ambientLights.size === 0) {
		return null;
	}
	let totalIntensity = 0;
	let accumR = 0;
	let accumG = 0;
	let accumB = 0;
	for (const light of ambientLights.values()) {
		totalIntensity += light.intensity;
		accumR += light.color[0] * light.intensity;
		accumG += light.color[1] * light.intensity;
		accumB += light.color[2] * light.intensity;
	}
	if (totalIntensity <= 0) {
		return null;
	}
	return {
		type: 'ambient',
		color: [
			accumR / totalIntensity,
			accumG / totalIntensity,
			accumB / totalIntensity,
		],
		intensity: totalIntensity,
	};
}

File: src/bmsx_cpp/render/shared/hardware/lighting.cpp
#include "render/shared/hardware/lighting.h"

#include <unordered_map>

namespace bmsx {
namespace {

std::vector<AmbientLight> s_ambientLights;
std::vector<DirectionalLight> s_directionalLights;
std::vector<PointLight> s_pointLights;
std::unordered_map<std::string, size_t> s_ambientLightIndices;
std::unordered_map<std::string, size_t> s_directionalLightIndices;
std::unordered_map<std::string, size_t> s_pointLightIndices;
bool s_hardwareLightingDirty = false;

template<typename Light>
void putHardwareLight(std::vector<Light>& lights, std::unordered_map<std::string, size_t>& indices, const std::string& id, const Light& light) {
	const auto it = indices.find(id);
	if (it == indices.end()) {
		indices[id] = lights.size();
		lights.push_back(light);
	} else {
		lights[it->second] = light;
	}
}

void markHardwareLightingDirty() {
	s_hardwareLightingDirty = true;
}

} // namespace

void putHardwareAmbientLight(const std::string& id, const AmbientLight& light) {
	putHardwareLight(s_ambientLights, s_ambientLightIndices, id, light);
	markHardwareLightingDirty();
}

void putHardwareDirectionalLight(const std::string& id, const DirectionalLight& light) {
	putHardwareLight(s_directionalLights, s_directionalLightIndices, id, light);
	markHardwareLightingDirty();
}

void putHardwarePointLight(const std::string& id, const PointLight& light) {
	putHardwareLight(s_pointLights, s_pointLightIndices, id, light);
	markHardwareLightingDirty();
}

void clearHardwareLighting() {
	const bool hadLights = !s_ambientLights.empty() || !s_directionalLights.empty() || !s_pointLights.empty();
	s_ambientLights.clear();
	s_directionalLights.clear();
	s_pointLights.clear();
	s_ambientLightIndices.clear();
	s_directionalLightIndices.clear();
	s_pointLightIndices.clear();
	if (hadLights) {
		s_hardwareLightingDirty = true;
	}
}

bool consumeHardwareLightingDirty() {
	if (!s_hardwareLightingDirty) {
		return false;
	}
	s_hardwareLightingDirty = false;
	return true;
}

const std::vector<AmbientLight>& getHardwareAmbientLights() {
	return s_ambientLights;
}

const std::vector<DirectionalLight>& getHardwareDirectionalLights() {
	return s_directionalLights;
}

const std::vector<PointLight>& getHardwarePointLights() {
	return s_pointLights;
}

std::optional<AmbientLight> resolveHardwareAmbientLight() {
	if (s_ambientLights.empty()) {
		return std::nullopt;
	}
	f32 totalIntensity = 0.0f;
	f32 accumR = 0.0f;
	f32 accumG = 0.0f;
	f32 accumB = 0.0f;
	for (const AmbientLight& light : s_ambientLights) {
		totalIntensity += light.intensity;
		accumR += light.color[0] * light.intensity;
		accumG += light.color[1] * light.intensity;
		accumB += light.color[2] * light.intensity;
	}
	if (totalIntensity <= 0.0f) {
		return std::nullopt;
	}
	return AmbientLight{
		{
			accumR / totalIntensity,
			accumG / totalIntensity,
			accumB / totalIntensity,
		},
		totalIntensity,
	};
}

} // namespace bmsx
